src/targets/gpu/lowering.cpp

```diff
@@ -229,6 +229,13 @@ struct miopen_apply
             for(auto& in : inputs)
             {
                 auto p_output = prog->insert_instruction(ret, hip_copy_from_gpu{}, in);
+                // the last copy back instruction needs a synchronization to ensure data copy
+                // completion
+                if(in == inputs.back())
+                {
+                    p_output = prog->insert_instruction(ret, hip_sync_device{}, p_output);
```

Comment:
This can be moved out of the loop. And `p_output` doesn't need to be passed in.

Reply:
Ah maybe `p_output` needs to be passed in so scheduling doesn't move it somewhere else. Of course, the scheduler could reorder these so the last input may not be the last copied. So it should take all copies as inputs.

Reply:
changed accordingly